src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java

```diff
@@ -425,80 +438,122 @@ public static CallSite metafactory(MethodHandles.Lookup caller,
+            if (bridgeCount < 0) {
+                throw new IllegalArgumentException("negative argument count");
+            }
+            if (bridgeCount > 0) {
+                bridges = extractArgs(args, argIndex, MethodType.class, bridgeCount);
+                argIndex += bridgeCount;
+            }
+        }
+        if (argIndex < args.length) {
+            throw new IllegalArgumentException("too many arguments");
         }
-        else
-            bridges = EMPTY_MT_ARRAY;
 
         boolean isSerializable = ((flags & FLAG_SERIALIZABLE) != 0);
         if (isSerializable) {
-            boolean foundSerializableSupertype = Serializable.class.isAssignableFrom(invokedType.returnType());
-            for (Class<?> c : markerInterfaces)
+            boolean foundSerializableSupertype = Serializable.class.isAssignableFrom(factoryType.returnType());
+            for (Class<?> c : interfaces)
                 foundSerializableSupertype |= Serializable.class.isAssignableFrom(c);
             if (!foundSerializableSupertype) {
-                markerInterfaces = Arrays.copyOf(markerInterfaces, markerInterfaces.length + 1);
-                markerInterfaces[markerInterfaces.length-1] = Serializable.class;
+                interfaces = Arrays.copyOf(interfaces, interfaces.length + 1);
+                interfaces[interfaces.length-1] = Serializable.class;
             }
         }
 
         AbstractValidatingLambdaMetafactory mf
-                = new InnerClassLambdaMetafactory(caller, invokedType,
-                                                  invokedName, samMethodType,
-                                                  implMethod,
-                                                  instantiatedMethodType,
+                = new InnerClassLambdaMetafactory(caller,
+                                                  factoryType,
+                                                  interfaceMethodName,
+                                                  interfaceMethodType,
+                                                  implementation,
+                                                  dynamicMethodType,
                                                   isSerializable,
-                                                  markerInterfaces, bridges);
+                                                  interfaces,
+                                                  bridges);
         mf.validateMetafactoryArgs();
         return mf.buildCallSite();
     }
+
+    private static <T> T extractArg(Object[] args, int index, Class<T> type) {
+        if (index >= args.length) {
+            throw new IllegalArgumentException("missing argument");
+        }
+        Object result = Objects.requireNonNull(args[index]);
+        if (!type.isInstance(result)) {
+            throw new IllegalArgumentException("argument has wrong type");
+        }
+        return type.cast(result);
```

Comment:
Can we just use a `return (T) result` as there will always be a checked cast on the caller's end, and this call seems redundant? The only shortcoming is that the call will raise an unchecked warning that needs to be suppressed. Or is this negligible after hotspot optimization?

Reply:
Eh, the scale here is quite small (never more than, say, 20 items), and an instanceof + a cast is a routine coding style that I'd expect to be optimized away. Doesn't seem worth the maintenance noise of a @SuppressWarnings.